fieldsolver/gridGlue.cpp

```diff
@@ -179,6 +189,117 @@ void feedMomentsIntoFsGrid(dccrg::Dccrg<SpatialCell,dccrg::Cartesian_Geometry>&
+  phiprof::start("BoxCar Filtering");
+
+  // Kernel Matrix-- Needs to be precomputed in memory and not recomputed at every call
+  Real kernel[3][3][3];
+
+  for (int i =0; i<3; i++){
+    for (int j =0; j<3; j++){
+      for (int k =0; k<3; k++){
+        
+        kernel[i][j][k]=1.0/27.0;  //normalized for 27 neighbours.
+        
+      }
+    }
+  }
+
+
+  phiprof::start("GhostUpdate");
+  momentsGrid.updateGhostCells(); //update ghost zones
+  phiprof::stop("GhostUpdate");
+
+  int stencilWidth = sizeof( kernel) / sizeof( kernel[0]); //get the stnecil width
+  int kernelOffset= stencilWidth/3; //this is the kernel offset -int
+  const int *mntDims= &momentsGrid.getLocalSize()[0];   //get local size for each proc.
+  FsGrid<std::array<Real, fsgrids::moments::N_MOMENTS>, 2> swapGrid = momentsGrid;  //swap array 
+  std::array<Real, fsgrids::moments::N_MOMENTS> *cell;  
+  std::array<Real,fsgrids::moments::N_MOMENTS> *swap;
+  const int maxRefLevel = mpiGrid.mapping.get_maximum_refinement_level();
+
+
+
+  for (int blurPass = 1; blurPass <= P::blurPasses; blurPass++){
+
+   // // Blurring Pass
+    phiprof::start("BlurPass");
+   // #pragma omp parallel for collapse(3)
+    for (int kk = 0; kk < mntDims[2]; kk++){
+      for (int jj = 0; jj < mntDims[1]; jj++){
+        for (int ii = 0; ii < mntDims[0]; ii++){
+         
+          // Dont filter if in max resolution region
+          int refLevel = technicalGrid.get(ii, jj, kk)->RefLevel;
+          if (refLevel == maxRefLevel || technicalGrid.get(ii, jj, kk)->sysBoundaryFlag == sysboundarytype::DO_NOT_COMPUTE)
+            {
+              continue;
+            }
+
+          for (int a=0; a<stencilWidth; a++){
+            for (int b=0; b<stencilWidth; b++){
+              for (int c=0; c<stencilWidth; c++){
+                   
+
+
+                int xn=ii+a-kernelOffset;
+                int yn=jj+b-kernelOffset;
+                int zn=kk+c-kernelOffset;
+
+                cell = momentsGrid.get(xn, yn, zn);
+                swap = swapGrid.get(xn, yn, zn);
+
+                for (int e = 0; e < fsgrids::moments::N_MOMENTS; ++e) {
+                    
+                  swap->at(e)+=cell->at(e) * kernel[a][b][c];
+
+
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+    phiprof::stop("BlurPass");
+
+    phiprof::start("GhostUpdate");
+    swapGrid.updateGhostCells();
+    phiprof::stop("GhostUpdate");
+
+    //  Copy swapGrid back to momentsGrid
+    phiprof::start("Swap->Moments");
+    // #pragma omp parallel for collapse(4)
+    for (int ii = 0; ii < mntDims[0]; ii++){
+      for (int jj = 0; jj < mntDims[1]; jj++){
+        for (int kk = 0; kk < mntDims[2]; kk++){
+          for (int e = 0; e < fsgrids::moments::N_MOMENTS; ++e){
+
+            // Get Arrays
+            cell = momentsGrid.get(ii, jj, kk);
+            swap = swapGrid.get(ii, jj, kk);
+
+            //Copy
+            cell->at(e) = swap->at(e);
```

Comment:
Really this could be a one-liner, less hassle for me to read but I don't mind if you prefer like this. The compiled code will be identical I assume.

Reply:
This has been requested many times I know, however the compiler doesn't let me swap pointers between the two of them. I must be doing something wrong I guess. I would also prefer a one liner here for readability purposes